### ISSUE TITLE
Developer Agent Training hangs

### ISSUE THREAD

**Issue by bogoconic1** [bogoconic1/Qgentic-AI#137]
Sometimes, the AI generated code, when executed went into a deadlock and the agent stalls until the script timeout is reached. We need a smarter way to handle this.

**Comment by bogoconic1** [comment]
So I think the approach is

Patch #115 so that the code execution is non-blocking (or maybe a background job) then have algorithms to poll for completion and check the logs for red flags. Automatically send a SIGKILL -9 if red flags are detected

**Comment by bogoconic1** [comment]
goal: solve this within 1.5 hours once started. no shipping crap. design or code quality can't suffer.

**Comment by bogoconic1** [comment]
## Proposed Solution: LLM Monitor Agent

Replace blocking `subprocess.run()` with `subprocess.Popen()` + an LLM agent that periodically inspects logs and system state, killing the process when it detects problems.

### How it works (concrete example)

**Setup**: Developer agent generates `train.py` and launches it via `Popen` with unbuffered output (`python -u`). Two reader threads stream stdout/stderr into a shared buffer. A monitoring loop polls every 1 second.

**Timeline of a NaN loss scenario:**

```
t=0s     Process starts. Reader threads begin streaming output.
t=5s     stdout: "Loading dataset..."
t=30s    stdout: "Dataset loaded. 50,000 samples."
t=35s    stdout: "Starting training with lr=0.01, batch_size=64"
t=40s    stdout: "Epoch 1/100 - loss: 2.345 - acc: 0.12"
t=90s    stdout: "Epoch 2/100 - loss: 1.876 - acc: 0.34"
t=120s   ⚡ MONITOR FIRES (120s interval reached)
```

The monitor agent receives:

```
<logs>
Loading dataset...
Dataset loaded. 50,000 samples.
Starting training with lr=0.01, batch_size=64
Epoch 1/100 - loss: 2.345 - acc: 0.12
Epoch 2/100 - loss: 1.876 - acc: 0.34
</logs>

seconds_since_last_output: 30
total_elapsed_seconds: 120
pid: 48291
```

LLM analyzes → loss is decreasing, training looks healthy → **returns `{"action": "continue", "reason": "Loss decreasing normally (2.345 → 1.876), training progressing as expected."}`**

```
t=140s   stdout: "Epoch 3/100 - loss: 1.102 - acc: 0.51"
t=190s   stdout: "Epoch 4/100 - loss: nan - acc: 0.00"
t=195s   stdout: "Epoch 5/100 - loss: nan - acc: 0.00"
t=200s   stdout: "Epoch 6/100 - loss: nan - acc: 0.00"
...
t=240s   ⚡ MONITOR FIRES
```

The monitor agent receives:

```
<logs>
...
Epoch 3/100 - loss: 1.102 - acc: 0.51
Epoch 4/100 - loss: nan - acc: 0.00
Epoch 5/100 - loss: nan - acc: 0.00
Epoch 6/100 - loss: nan - acc: 0.00
Epoch 7/100 - loss: nan - acc: 0.00
</logs>

seconds_since_last_output: 2
total_elapsed_seconds: 240
pid: 48291
```

LLM detects NaN loss. Before making its decision, it calls its `execute_bash` tool:

```bash
→ nvidia-smi --query-gpu=utilization.gpu,memory.used --format=csv,noheader
  98%, 15234 MiB    # GPU is being used but wasted on NaN computation
```

LLM returns: **`{"action": "kill", "reason": "Training diverged: loss became NaN at epoch 4 (was 1.102 at epoch 3). GPU still at 98% utilization computing NaN gradients. Learning rate 0.01 may be too high. Recommend reducing lr or adding gradient clipping."}`**

```
t=240s   🔪 SIGKILL sent to process group (kills train.py + all DataLoader workers)
```

The developer agent receives:
```
Code execution killed by log monitor: Training diverged: loss became NaN at epoch 4
(was 1.102 at epoch 3). GPU still at 98% utilization computing NaN gradients.
Learning rate 0.01 may be too high. Recommend reducing lr or adding gradient clipping.

Last output:
Epoch 5/100 - loss: nan - acc: 0.00
Epoch 6/100 - loss: nan - acc: 0.00
Epoch 7/100 - loss: nan - acc: 0.00
```

The developer agent then generates a fixed `train.py` with gradient clipping and a lower learning rate.

**Without this fix**: the agent would have waited the full **12-hour timeout** before discovering the NaN issue. With the fix: **4 minutes**.

### What the monitor catches

The LLM reasons freely — no hardcoded rules. Examples:
- **NaN/Inf loss** → kill (training is unrecoverable)
- **Loss plateau for 30+ epochs** → kill (wasting compute)
- **Deadlock** (no output + 0% GPU via `nvidia-smi`) → kill
- **Downloading large dataset** (no output but expected) → continue
- **Repeated identical lines** (infinite loop) → kill
- **CUDA OOM warnings** → kill before full crash
- **Slow but progressing training** → continue

### Architecture

- `execute_code()` in `tools/developer.py` — Popen + monitoring loop
- `monitor_logs()` — tool-calling LLM agent (same pattern as SOTA suggestions loop)
- `execute_bash` tool — lets the LLM run diagnostic commands (`nvidia-smi`, `ps`, `free`, etc.)
- Max 3 tool steps per monitor call to keep overhead low
- Configurable interval (`log_monitor_interval: 120` in config.yaml)

Closes #137, closes #142.
